### PR TITLE
867 devices in blueapi context

### PR DESCRIPTION
**Pull request by Tom-Willemsen**
Fixes https://github.com/DiamondLightSource/hyperion/issues/867

This refactors our hyperion plans to no longer refer to things like `i03.backlight()` directly, and instead to pull devices out of a `blueapi` context.

The main approach at present is that each class defines a composite of devices it needs (as a dataclass), in it's setup method.

as an example, something like:

```python
@dataclasses.dataclass
class SomePlanComposite():
    backlight: Backlight
    eiger: Eiger
```

In the plan's `create_devices`, this composite is then instantiated with the appropriate devices from the `blueapi` context. This composite instance is then passed as one of the arguments to the plan, and relevant devices (or the entire composite) passed down into sub-plans as necessary.

I've tried to design this so that it should be relatively easy to move to blueapi's `inject` mechanism when we do https://github.com/DiamondLightSource/hyperion/issues/868 (might need a thin shim), as now all of the dependencies of each plan are passed around as explicit arguments rather than implicit and scattered throughout sub-plans.

I've tried to keep the API type-safe for developers where possible, as before, which is why I chose to use a dataclasses in the above format rather than passing around e.g. a `Dict[str, Device]`. This also lets us check we got a device of the expected type out of the context.

Some API decisions which I'm not necessarily sure about:
- The handling of `wait_for_connection=False` is a bit hacky, it's currently "known about" in two places, one when you setup the context (which honors the wait_for_connection setting) and one when the device composite is actually created (at which point you're starting a plan and therefore always need the devices to be connected).
  * I think this will likely fall out of https://github.com/DiamondLightSource/hyperion/issues/868 once that issue is done, as  `blueapi` will have to be responsible for making sure that any devices it injects are actually connected at the point they're used
  * Unless we want to make the plan itself responsible for this? But that doesn't seem right.
  * Happy to hear any other alternative approaches
- Post-init in `device_instantiation` (for example - loading aperture positions). At the point when we're building the `context` we don't know which plan we'll be running but still need to set up the devices. Currently I've moved loading aperture positions into a `__post_init__` on the dataclasses described above, so that this logic happens when the relevant device is about to be used. 
  * This will likely change again when we do https://github.com/DiamondLightSource/hyperion/issues/868 as `blueapi` will then somehow have to be told about any post-device-creation initialisation that needs to be done.
  * But again, happy to hear alternative approaches on this...

### To test
- Ensure that individual plans do not reference `i03.<anything>` any more.
  * Unit tests might still reference i03, in particular mocks set up using things like `mock_backlight = i03.backlight(fake_with_ophyd_sim=True)`. This seems ok to me as it's just for simulated devices.
- No *new* s03 test failures. Some tests fail on `main` and still fail on this branch (e.g. rotation scans, fgs).
- **Will need testing on beamline(s).** This is a change that touches all plans and isn't possible to test fully from a dev machine as far as I know.

### CODE REVIEW

**Comment by callumforrester** [comment]
Unsure if it's a multi-step process, but I think the eventual aim should be to remove the experiment registry entirely and just replace it with a hierarchy of plans. I.e. there is a top level plan called `fast_grid_scan_experiment` which takes the correct parameters, sets up the subscriptions via `yield from subscribe(...)`, and calls the actual `fast_grid_scan` plan. 

The philosophy here is that blueapi can be a nice, stable, standardised core that's the same for everyone, and per-beamline customization is kept in the plans as far as is reasonable. This is a very core-ivory-tower point of view though and hasn't yet been tested in anger, so I welcome differing opinions @DominicOram and @Tom-Willemsen. We deliberately left this part of the design loose in anticipation of feedback.

**Comment by Tom-Willemsen** [comment]
> Unsure if it's a multi-step process, but I think the eventual aim should be to remove the experiment registry entirely and just replace it with a hierarchy of plans. I.e. there is a top level plan called `fast_grid_scan_experiment` which takes the correct parameters, sets up the subscriptions via `yield from subscribe(...)`, and calls the actual `fast_grid_scan` plan.
> 
> The philosophy here is that blueapi can be a nice, stable, standardised core that's the same for everyone, and per-beamline customization is kept in the plans as far as is reasonable. This is a very core-ivory-tower point of view though and hasn't yet been tested in anger, so I welcome differing opinions @DominicOram and @Tom-Willemsen. We deliberately left this part of the design loose in anticipation of feedback.

I agree with what you've written, and I'm hoping that this PR doesn't conflict too much with that view. The only additional coupling to `experiment_registry` in this PR is that the object returned by `setup` (in practice, always a composite) is passed as the first argument to the plan. In my mind the way to move this out of the experiment registry level would be something along the lines of:

```python
from blueapi import inject_composite

@dataclasses.dataclass
class MyPlanComposite:
    backlight: Backlight
    eiger: Eiger
    ...

def my_experiment_plan(
        composite: MyPlanComposite = inject_composite(MyPlanComposite),
        ...
    ):
    yield from prepare(composite)
    yield from do_it(composite)
    yield from clean_up(composite)
```

where `inject_composite` is similar to the existing blueapi `inject` mechanism, but instead of injecting a *single* device at a time, knows how to iterate through the members of the referenced composite and inject a "populated" composite. Essentially moving the logic from `device_composite_from_context` introduced in this PR into blueapi itself. Exact syntax can be argued/refined...

The main thing I was trying to avoid was having to write something very drawn out like:

```python
from blueapi import inject

def my_experiment_plan(
        backlight: Backlight = inject("backlight"),
        eiger: Eiger = inject("eiger"),
        ...<20 more devices>...
    ):
    yield from prepare(eiger, backlight, <20 more devices>)
    yield from do_it(eiger, backlight, <20 more devices>)
    yield from clean_up(eiger, backlight, <20 more devices>)
```
in every single plan (and passing through lots of devices into subplans too).

I think the *option* to inject single device(s) should remain though, for simpler plans that only need a limited number of devices.

---

This PR is indeed just a first step though - moving away from hard-coded `i03.something()` to something generic across beamlines. There are further issues open in this repo to move to blueapi's management components, which would then mean using something more like `blueapi`'s `inject` mechanism or similar.

---

Does that sound sensible?

**Comment by DominicOram** [comment]
> Unsure if it's a multi-step process, but I think the eventual aim should be to remove the experiment registry entirely and just replace it with a hierarchy of plans. I.e. there is a top level plan called `fast_grid_scan_experiment` which takes the correct parameters, sets up the subscriptions via `yield from subscribe(...)`, and calls the actual `fast_grid_scan` plan.
> 
> The philosophy here is that blueapi can be a nice, stable, standardised core that's the same for everyone, and per-beamline customization is kept in the plans as far as is reasonable. This is a very core-ivory-tower point of view though and hasn't yet been tested in anger, so I welcome differing opinions @DominicOram and @Tom-Willemsen. We deliberately left this part of the design loose in anticipation of feedback.

Yes, this was my understanding of the end goal. This is a step towards that to try and work incrementally to where we want to be. I think the next issue is probably https://github.com/DiamondLightSource/hyperion/issues/868 then I think a full removal of `experiment_registry` and clean up

**Comment by callumforrester** [comment]
@DominicOram and @Tom-Willemsen, that all sounds good to me! Will let you know if I have any other thoughts.

File: src/hyperion/experiment_plans/grid_detect_then_xray_centre_plan.py [hunk 1]
```diff
@@ -43,18 +54,33 @@
...
 
-def create_devices():
-    fgs_create_devices()
-    oav_create_devices()
-
-    aperture_positions = AperturePositions.from_gda_beamline_params(
-        get_beamline_parameters()
-    )
+@dataclasses.dataclass
+class GridDetectThenXRayCentreComposite:
```
[the remark this answers is not in the document]
Comment (DominicOram) [reply src/hyperion/experiment_plans/grid_detect_then_xray_centre_plan.py:58]: I don't want this PR to be blocked based on the long discussion of this though so happy to leave as is for now

File: src/hyperion/experiment_plans/flyscan_xray_centre_plan.py [hunk 2]
```diff
@@ -39,63 +35,50 @@
...
-        self.synchrotron: Undulator = i03.synchrotron(fake_with_ophyd_sim=fake)
-        self.zebra: Zebra = i03.zebra(fake_with_ophyd_sim=fake)
-        self.attenuator: Attenuator = i03.attenuator(fake_with_ophyd_sim=fake)
-
-
-flyscan_xray_centre_composite: GridscanComposite | None = None
 
 
-def create_devices():
+def create_devices(context: BlueskyContext) -> FlyScanXRayCentreComposite:
```
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply src/hyperion/experiment_plans/flyscan_xray_centre_plan.py:79]: I started doing this, but ran into a bunch of fiddly stuff with mocking in the system tests. Given that it'll be refactored shortly anyway I decided not to sink too much effort into it and have just left as-is.

File: src/hyperion/utils/context.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+class _IsDataclass(Protocol):
+    """Protocol followed by any dataclass"""
+
+    __dataclass_fields__: ClassVar[Dict]
+
+
+DT = TypeVar("DT", bound=_IsDataclass)
+
+
+def find_device_in_context(
```
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply src/hyperion/utils/context.py:31]: https://github.com/DiamondLightSource/blueapi/issues/309

File: src/hyperion/utils/context.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    # Ideally would use context.with_dodal_module, but it doesn't support
+    # passing through wait_for_connection or fake_with_ophyd_sim
+    # See https://github.com/DiamondLightSource/blueapi/pull/304
+    for name, device in make_all_devices(
+        _get_beamline_specific_device_module(),
+        wait_for_connection=wait_for_connection,
+        fake_with_ophyd_sim=fake_with_ophyd_sim,
+    ).items():
+        context.device(device, name=name)
```
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply src/hyperion/utils/context.py:124]: This possibly interacts a bit with the subject of tomorrow's meeting; I'll wait until after that meeting to see what we decide and where to go with this.

File: src/hyperion/utils/context.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        # At the point where we're actually making a device composite, i.e. starting a plan with these devices,
+        # we need all the referenced devices to be connected.
+        _wait_for_connection(device=device)
+
+        devices[field.name] = device
+
+    return dc(**devices)
+
+
+def _get_beamline_specific_device_module() -> ModuleType:
```
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply src/hyperion/utils/context.py:78]: https://github.com/DiamondLightSource/dodal/pull/166

**Review by DominicOram — CHANGES_REQUESTED** [review]
I think we can also now remove both `get_beamline_prefixes`and associated tests as we're doing this in `dodal`

File: src/hyperion/utils/context.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (DominicOram) [reply src/hyperion/utils/context.py:78]: Now merged, please update here

File: src/hyperion/utils/context.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (DominicOram) [reply src/hyperion/utils/context.py]: I still think we should merge https://github.com/DiamondLightSource/blueapi/pull/304 and remove here

File: src/hyperion/experiment_plans/tests/test_rotation_scan_plan.py [hunk 6]
```diff
@@ -108,29 +111,32 @@ def run_full_rotation_plan(
     undulator: Undulator,
     flux: Flux,
 ):
+    devices = RotationScanComposite(
```
[the remark this answers is not in the document]
Comment (DominicOram) [reply src/hyperion/experiment_plans/tests/test_rotation_scan_plan.py]: Should: Can't we use this for the other composites created in this file?

File: setup.cfg [hunk 7]
```diff
@@ -19,7 +19,7 @@ package_dir =
 install_requires =
     bluesky
     pyepics
-    blueapi
+    blueapi @ git+https://github.com/Tom-Willemsen/blueapi@unpin_dodal
```
[the remark this answers is not in the document]
Comment (DominicOram) [reply setup.cfg:22]: Hopefully https://github.com/DiamondLightSource/blueapi/pull/312 sorts this

File: src/hyperion/experiment_plans/grid_detect_then_xray_centre_plan.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply src/hyperion/experiment_plans/grid_detect_then_xray_centre_plan.py:62]: I think we had a discussion and both weren't sure what the best approach was. Let's try with it as it is at the moment, and we can revisit this decision if it becomes cumbersome. I'm just keen to get *something* through on this PR even if it's not necessarily perfect for all possible use cases.

File: setup.cfg (hunk 7 above)
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply setup.cfg]: Should be picked up as soon as https://github.com/DiamondLightSource/blueapi/pull/312 is merged I think.

**Review by DominicOram — APPROVED** [review]
Great, thank you Tom

**Pull request merged**